Zend/zend_attributes.c

```diff
@@ -35,31 +35,41 @@ static zend_object_handlers attributes_object_handlers_sensitive_parameter_value
 static HashTable internal_attributes;
 
 void validate_attribute(zend_attribute *attr, uint32_t target, zend_class_entry *scope)
+{
+}
+
+ZEND_API uint32_t zend_attribute_attribute_get_flags(zend_attribute *attr)
 {
 	// TODO: More proper signature validation: Too many args, incorrect arg names.
 	if (attr->argc > 0) {
 		zval flags;
 
-		/* As this is run in the middle of compilation, fetch the attribute value without
-		 * specifying a scope. The class is not fully linked yet, and we may seen an
-		 * inconsistent state. */
 		if (FAILURE == zend_get_attribute_value(&flags, attr, 0, NULL)) {
```

Comment:
Should you specify the scope now? The previous reflection implementation did. My test case appears to pass even without, but maybe I did something wrong:

```
<?php

class Foo {
	protected const TARGET = Attribute::TARGET_CLASS | Attribute::TARGET_PARAMETER;
}

#[\Attribute(parent::TARGET)]
class Bar extends Foo {

}

$r = new ReflectionClass('Bar');
var_dump($r->getAttributes()[0]->newInstance());
```

Reply:
The following doesn't work.

```php
class Foo {
	protected const TARGET = Attribute::TARGET_CLASS | Attribute::TARGET_PARAMETER;
}

#[\Attribute(parent::TARGET)]
class Bar extends Foo {}

#[\Bar]
class Baz {}

$r = new ReflectionClass('Baz');
var_dump($r->getAttributes()[0]->newInstance());
```

Your example avoids the code path from this PR completely.

Whether this _should_ work is a bit philosophical. Does `self`/`parent` refer to the class, given it is _outside_ the class / attached _to_ the class? Either way, looking at the other instances of `zend_get_attribute_value`, scope is always passed, so let's do the same here.

Reply:
> Whether this _should_ work is a bit philosophical. Does `self`/`parent` refer to the class, given it is _outside_ the class / attached _to_ the class? Either way, looking at the other instances of `zend_get_attribute_value`, scope is always passed, so let's do the same here.

Yeah, I stumbled about the same question for `#[\Deprecated]`, wondering what I need to pass as the scope: The attribute class (i.e. the CE for Deprecated) or the class of the method having the attribute.

The result of my testing with userland attributes was this: https://github.com/php/php-src/blob/7f4850b1e20fd3bd3e1c939e8d886c718389d50d/Zend/tests/attributes/deprecated/functions/017.phpt